server/src/snapshot.rs

```diff
@@ -0,0 +1,430 @@
+        status.meta_written = true;
+    }
+
+    pub fn finished(&self) -> bool {
+        let status = self.status.lock().expect("mutex poisoned");
+
+        for s in &status.table_states {
+            match s {
+                TableState::Finished => continue,
+                _ => return false,
+            }
+        }
+
+        true
+    }
+
+    fn should_stop(&self) -> bool {
+        let status = self.status.lock().expect("mutex poisoned");
+        status.stop_on_next_update
+    }
+
+    async fn run(&self, notify: Option<oneshot::Sender<()>>) -> Result<()> {
+        while let Some((pos, table_name)) = self.next_table() {
+            let batch = self
+                .partition
+                .table_to_arrow(table_name, &[])
+                .map_err(|e| Box::new(e) as _)
+                .context(PartitionError)?;
+
+            let file_name = format!("{}/{}.parquet", &self.data_path, table_name);
+
+            self.write_batch(batch, &file_name).await?;
+            self.mark_table_finished(pos);
+
+            if self.should_stop() {
+                return StoppedEarly.fail();
+            }
+        }
+
+        let partition_meta_path =
+            format!("{}/{}.json", &self.metadata_path, &self.partition_meta.key);
+        let json_data = serde_json::to_vec(&self.partition_meta).context(JsonGenerationError)?;
+        let data = Bytes::from(json_data);
+        let len = data.len();
+        let stream_data = std::io::Result::Ok(data);
+        self.store
+            .put(
+                &partition_meta_path,
+                futures::stream::once(async move { stream_data }),
+                len,
+            )
+            .await
+            .context(WritingToObjectStore)?;
+
+        self.mark_meta_written();
+
+        if let Some(notify) = notify {
+            if let Err(e) = notify.send(()) {
+                error!("error sending notify: {:?}", e);
+            }
+        }
+
+        Ok(())
+    }
+
+    async fn write_batch(&self, batch: RecordBatch, file_name: &str) -> Result<()> {
+        let mem_writer = MemWriter::default();
+        {
+            let mut writer = ArrowWriter::try_new(mem_writer.clone(), batch.schema(), None)
+                .context(OpeningParquetWriter)?;
+            writer.write(&batch).context(WritingParquetToMemory)?;
+            writer.close().context(ClosingParquetWriter)?;
+        } // drop the reference to the MemWriter that the SerializedFileWriter has
+
+        let data = mem_writer
+            .into_inner()
+            .expect("Nothing else should have a reference here");
+
+        let len = data.len();
+        let data = Bytes::from(data);
+        let stream_data = std::io::Result::Ok(data);
+
+        self.store
+            .put(
+                &file_name,
+                futures::stream::once(async move { stream_data }),
+                len,
+            )
+            .await
+            .context(WritingToObjectStore)?;
+
+        Ok(())
+    }
```

Comment:
You can just return the `put` call.

```suggestion
            .await
            .context(WritingToObjectStore)
    }
```